notes.txt

```diff
@@ -0,0 +1,241 @@
+
+# process all unprocessed records for each host zone
+for record in records where record.type == REWARD:
+  def bankSendAck(ack):
+    record = getRecord(ack.id)
+    record.type = DEPOSIT
+    record.status = STAKE
+  # do upon completion
+  ICABankSend.Ack = bankSendAck
+  ICABankSend(from=withdrawal, to=delegat, amount=record.amount)
+
+# add new records for rewards that accrued in the previous epoch
+for hz in getAllHostZones():
+  def addRewardsRecord(hz):
+    # define a query callback function
+    def queryCB(hz):
+      recordBalance = sumAllRewardsRecords()
+      newBalance = balance - recordBalance
+      createNewRewardsRecord(newBalance)
+    wA = hz.withdrawalAccount()
+    query(address=wA, queryType=balance, cb=queryCB)
+
+
+# NOTE: this is executed after processing records because although ICA calls
+# _are_ ordered (meaning we could first send balances from withdrawal -> delegate account),
+# there's no way to enforce that the records are updated on stride without chaining these.
+# unclear whether chaining is a robust approach. What happens if one step fails?
+# what do we have to roll back?
+# easier to think about these in isolation IMO
+1. ICQ wA balance
+2. ICQ CB: create records, BankSend ICA call them from withdrawal -> delegate account
+3. Bank send ICA Ack: update deposit records, iterate through them and MsgDelegate ICA call them
+4. MsgDelegate ICA ack: delete deposit records
+    
+
+----------
+NEXT EPOCH START
+...
+
+
+
+
+
+
+
+
+
+*******************************************************************
+******************************UNBONDING****************************
+*******************************************************************
+
+# SIMPLIFYING ASSUMPTION: Only allow users to process 1 unbonding / block
+# this allows us to more easily track unique unbonding records for a user
+# by keying them as senderAddress:hostZoneId:amount:block
+
+
+# SECTION 3
+# 0 Add UnbondingRecord type
+# 1 Add RedeemStake logic outlined below
+# 2 Update Epoch end to
+  # a. PROCESS RECORDS - ENQUEUE
+  # b. CACHE MOST RECENT BLOCK HEIGHT FOR HOST ZONE
+  # c. PROCESS RECORDS - SEND BACK TO USERS
+
+
+>>> EPOCH START
+----------
+
+# users call redeem stake
+def redeemStake(tokens):
+  unbondingRecord = newUnbondingRecord()
+  # getUnbondingBlockHeight queries the unbonding record
+  # in the creation ack and adds a dequeue unbonding height
+  # if the query fails, it's kept in the pending queue
+  # which we issue queries for, over and over
+  unbondingRecord.blockHeight = getUnbondingBlockHeight()
+  unprocessedUnbondingRecords.enqueue(unbondingRecord)
+
+>>> EPOCH END
+
+# PROCESS RECORDS - ENQUEUE
+for record in unprocessedRecords:
+  # query unbonding height
+  # in ack
+    # fetch record
+    # delete from unprocessed records
+    # add unbonding height
+    # add to processed records
+
+# CACHE MOST RECENT BLOCK HEIGHT FOR HOST ZONE
+for zone in hostZones:
+  client = getClient
+  mostRecentBlock = client.height
```

Comment:
again, might need to port this to clock time. hoping we can do it with block height. if we do it by block height, we prob wanna trigger a LC update before this line

Reply:
I imagine it might actually be easier to do by clock time? Clock time should be consistent across Stride / other zones, so we wouldn't even have to update the LC! (might be missing something here though, don't have a super solid understanding of how clock time works in Cosmos)